### DISCUSSION TITLE
Island renderer architecture vs nested hydrated components

### DISCUSSION THREAD

**Discussion by innermatrix** [withastro/roadmap#18] [Proposal]
This started as [an issue](https://github.com/withastro/astro/issues/2016) that was deemed not to be a bug and closed, but I disagree with this behavior and want to bring it up again in hope of arriving at a consensus about how this use case should behave.

## What I am trying to do

1. Have a hydrated component that contains another hydrated component *with intervening non-hydrated components*
2. Have the ancestor and the descendant hydrated components share context

When I say "share context", I am specifically referring to the kind of behavior that is provided by context APIs in [React](https://reactjs.org/docs/hooks-reference.html#usecontext) / [Preact](https://preactjs.com/guide/v10/context/) / [Svelte](https://svelte.dev/tutorial/context-api), in which an ancestor component can set a variable in a local context, and a descendant component can access that variable in its local context. This is different from state APIs (such as provided by `nanostores`), which allow inter-component communication that is agnostic of component containment hierarchy. 

I am assuming here that the contributors to this discussion are already aware of the reasons for why both context-based APIs (such as `getContext`) and state-based APIs (such as `nanostores`) exist, and that it is understood that these two types of APIs do not meet the same needs. In other words, "use state APIs" does not address the issue raised here. 

Here's an example of Astro code that I would like to work that currently does not:

```
<div>
  <Component1 client:only>
    <div>
      <Component2 client:only />
    </div>
  </Component1>
</div>
```

Specifically, if `Component1` declares a context (using the framework-specific context API) and `Component2` uses that context (ditto), what we find is that the context is not shared between the two components, and as a result hierarchy-based state management is impossible.

What does work in this simple case is if the Astro code is changed to 

```
<div>
  <Component1 client:only />
</div>
```

and `Component1.tsx` is changed to

```
<div>
  <Component2 client:only />
</div>
```

However, if the inner `div` had been an Astro component instead:

```
<div>
  <Component1 client:only>
    <AstroComponent>
      <Component2 client:only />
    </AstroComponent>
  </Component1>
</div>
```

then I can't think of anything that will make this work, because moving `<AstroComponent>` inside `Component1.tsx` would require the Astro component to be available on the client. 

The net result of this is that if a page contains two components (here `Component1` and `Component2`) that are using the same framework renderer, and one of them contains the other, then they will only work correctly if the entire hierarchy contained by the ancestor component is moved into the framework renderer, and contains no Astro components. What this leads to is that the only hydrated components that are actually usable are the ones that don't use slots/children; as soon as a hydrated component uses slots/children, it is possible to pass it slots/children that will not work correctly (specifically, any slot/child that contains another hydrated component). 

Beyond "I don't think island architecture should break context-based data sharing", I don't have a solution here. Yet? I am hoping that the discussion here will point in the direction of a viable approach to solving this problem. If/when it does, I am happy to work on turning it into an RFC.

**Comment by FredKSchott** [comment]
Thanks @innermatrix! I agree that this is the best place to discuss. I think @natemoo-re is probably the best person to get feedback from here.

I think there is definitely general interest in getting some shared context solution figured out for React and Vue. I know that this is blocking a few popular packages that don't work with Astro because they assume a context is shared across the entire website as a single SPA application. 

We've talked about being able to set a single context that is shared across all islands. Would that also solve this problem? Setting context per island feels related but may not be what most users want.

Personally: I'm skeptical of nesting client components in Astro syntax, since I don't think we can reliably mimic how nesting should work in that specific component framework. Maybe we should push to revisit this as an API decision for v1.0?

**Comment by innermatrix** [comment]
I don't think that a single global context addresses this issue; the main use case for contexts is in situations when the same container component appears multiple times on the same page. As in:

```
<Component1>
  <stuff>
    <Component2/>
  </stuff>
</Component1>
<Component1>
  <stuff>
    <Component2/>
  </stuff>
</Component1>
```

In this situation, the different instances of `Component1` each have their own context instance, and the different instances of `Component2` can behave differently based on which `Component1` they are in. I don't see how a single global context would address that. 

I am not convinced that we can't attain the framework nesting behavior, but admittedly I haven't spent as much time thinking about this as others have. If we think that this is a desirable direction, though, maybe I should 🙂

**Reply by innermatrix** [reply]
Also @FredKSchott the need for a context API has been recognized at the WebComponents CG level and there is now a [draft proposal](https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/context.md) for a cross-framework context API. 

I would highly recommend reading [the discussion](https://github.com/webcomponents-cg/community-protocols/issues/2) that lead to this draft, as the cross-framework goals of WCCG are related to the goals of Astro, and I think that the rationale behind the draft proposal speaks directly to some issues we have touched on in this thread.

**Reply by matthewp** [reply]
Can you create a codepen that shows this working even in the client? I'd love to see this example working, nothing that `Component1` and `Component2` exist as separate `render(<Component />, root)` entrypoints.

**Comment by vzam** [comment]
I have found this proposal after having similar issues with the context using vue and would like to raise it again for discussion. I am aware of the workaround but it doesn't apply in my case, because I need the `AstroComponent` wrapper for `Component2` due to SSR. The data is pretty heavy and I really want to avoid having to map it on the client. Instead, I want astro to generate the components server-side and have them rendered on the client in the context of `Component1`.

The previously suggested solution of explicitly setting the context on the astro islands would be sufficient for my use-case, but I can also see why the approach with sharing state throughout descendants of the same framework is appealing.

I am wondering about this issue being forgotten, because it makes components that heavily depend on the context somewhat difficult to use in combination with SSR. Are there any extensive documents that I could read futher to get some info on why this issue is not being addressed?

**Comment by predaytor** [comment]
Inability to share framework specific Context API is the main issue for me using Astro. We can create a wrapper component like `SubscribeFormDialog.tsx` that encapsulates all the logic, a nested tree of components `DialogRoot`, `DialogTrigger` etc. and it will work fine until we need to pass some complex structured data from the server.

For example, some children require a special "slot" component rather than passing a single descendant. It is possible to create such a component tree (which will be rendered server-side), but we need to be able to share the context in the hydration tree, which is currently not possible. I believe that in this case only the parent component could be hydrated.
```astro
---
import { Image } from 'astro:assets';

import * as Carousel from './Carousel';
import ArrowLeftIcon from './ArrowLeftIcon';
import ArrowRightIcon from './ArrowRightIcon';
---

<div class="projects">
	{
		projects.map(({ gallery }) => (
			<Carousel.Root client:idle>
				<Carousel.Navigation client:idle>
					<Carousel.Previous client:idle>
						<ArrowLeftIcon aria-hidden="true" />
					</Carousel.Previous>

					<Carousel.Next client:idle>
						<ArrowRightIcon aria-hidden="true" />
					</Carousel.Next>
				</Carousel.Navigation>

				<Carousel.Scroller client:idle>
					{gallery.map(slide => {
						const {
							sourceUrl,
							mediaDetails: { width, height },
							altText,
						} = slide;

						return (
							<Carousel.Slide client:idle>
								<Image
									src={sourceUrl}
									width={width}
									height={height}
									alt={altText}
								/>
							</Carousel.Slide>
						);
					})}
				</Carousel.Scroller>
			</Carousel.Root>
		))
	}
</div>
```
